### ISSUE TITLE
Stub session delays actual session

### ISSUE THREAD

**Issue by nicolasbouffard** [webdriverio/webdriverio#6191]
Stub session delays actual session

**Environment (please complete the following information):**
 - **WebdriverIO version:** 6.4.6
 - **Mode:** WDIO Testrunner
 - **If WDIO Testrunner, running sync/async:** sync
 - **Node.js version:** 12.18.4
 - **NPM version:** 6.14.6
 - **Browser name and version:** Any Android app
 - **Platform name and version:** Ubuntu 20.04 LTS
 - **Additional wdio packages used (if applicable):** @wdio/mocha-framework

**Config of WebdriverIO**
Here's a summary of our WebdriverIO config :
```js
{
  runner: "local",
  path: "/wd/hub",
  specs: ["./test/**/*.js"],
  exclude: [],
  maxInstances: 10,
  maxInstancesPerCapability: 10,
  execArgv: [],
  logLevel: "info",
  logLevels: {
    webdriver: "info",
    "@wdio/applitools-service": "info"
  },
  bail: 0,
  baseUrl: "http://localhost",
  waitforTimeout: 20000,
  filesToWatch: [],
  connectionRetryTimeout: 60000,
  connectionRetryCount: 3,
  framework: "mocha",
  reporters: [
    [
      // We use a custom reporter, which isn't relevant to the issue I think.
    ]
  ],
  mochaOpts: {
    ui: "bdd",
    timeout: 600000,
    bail: false,
    compilers: ["@babel/register"],
    require: [
      "chai/register-should",
      "chai/register-assert",
    ]
  },
}
```

And here's how we use the runner :

```js
const wdio = new Launcher(`path/to/wdio.shared.conf.js`, {
  specs: ["path/to/our/script.js"],
  hostname: "appium-server-hostname",
  port: 4723,
  capabilities: [{
    // Basic smartphone testing capabilities that are irrelevant to the issue I think.
  }]
});
wdio.run().then(
  code => {
    process.exit(code);
  },
  error => {
    console.error("Launcher failed to start the test", error.stacktrace);
    process.exit(1);
  }
);
```

**Describe the bug**
WebdriverIO seems to initiate a "stub" session **before** the actual "webdriver" session, as shown in the logs below. In many cases, this stub session is immediately followed by the actual session, so it's not really an issue. But in our main use case, it takes sowhere in the neighborhood of 35s to complete, delaying our whole test suite.

**To Reproduce**
I can't provide anything else to reproduce easily right now, because the issue occurs in a complex environment with proprietary code. Besides, this seems to be solely caused by WebdriverIO, which takes too much time doing this `protocol-stub` thing, so I don't think more details about our environment would help much.

**Expected behavior**
We expect our test suite to begin right away as we don't understand what this stub session is, and as it doesn't seem to do anything. Actually, by messing with ![this part](https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-runner/src/index.js#L62), we managed to skip the `protocol-stub` session altogether, therefore avoiding any delay, and without any impact in our suite. At the very least, we expect WebdriverIO to not delay the actual session's initiation this much.

**Log**
(I know we're expected to create gists, but since our issue occurs at the very beginning, I didn't feel like it was necessary to include logs beyond that point and, therefore, to create a gist just for 10 lines of log)
As you can see, the "stub" session initiates at 13:26:08 and the actual session at 13:26:43, 35s later.

```
Execution of 1 spec files started at 2020-12-02T13:26:05.132Z

2020-12-02T13:26:05.152Z INFO @wdio/cli:launcher: Run onPrepare hook
2020-12-02T13:26:05.166Z INFO @wdio/cli:launcher: Run onWorkerStart hook
2020-12-02T13:26:05.170Z INFO @wdio/local-runner: Start worker 0-0 with arg: 
[0-0] 2020-12-02T13:26:07.839Z INFO @wdio/local-runner: Run worker command: run
[0-0] 2020-12-02T13:26:08.236Z INFO webdriverio: Initiate new session using the ./protocol-stub protocol
[0-0] RUNNING in com.some.app - script.js
[0-0] 2020-12-02T13:26:43.333Z INFO webdriverio: Initiate new session using the webdriver protocol
[0-0] 2020-12-02T13:26:43.345Z INFO webdriver: [POST] http://appium-server-hostname:4723/wd/hub/session
```

**Comment by mgrybyk** [comment]
Without steps to reproduce it can't be fixed. Also, I'm pretty much sure stub session has nothing to do with the delay you see.
Do you use any services? Do you run any code in wdio hooks?

**Comment by christian-bromann** [comment]
@mgrybyk might be worth thinking about adding some debug logs to time the service hook execution.

**Comment by nicolasbouffard** [comment]
> Do you use any services? Do you run any code in wdio hooks?

We don't use any services, but we do use hooks. I already disabled all of them however in an attempt to find who was the culprit, and it changed nothing.

After doing some more testing, I managed to narrow down the issue to the use of Babel. Indeed, by simply commenting out the `compilers: ["@babel/register"],` line in `mochaOpts`, the delay between the stub and webdriver sessions disappears. The hardware that suffers from the delay is a [PC engine](https://www.pcengines.ch/apu2.htm), so I thought this could be the issue, but running Babel shouldn't require an 8 core CPU, especially for handling scripts that are only ~20 lines long.

I tried to add some ES6 syntax (e.g.: arrow functions) to my script to confirm that, without Babel, we won't be able to use ES6 in our scripts. But to my surprise : it still works.

Afer reading again [the documentation section about Babel](https://webdriver.io/docs/babel.html), I decided to remove `@babel/register` entirely from my dependencies, in case it would be "automatically used" if present, kind of like "@wdio/sync" turns on sync mode all by itself. But again : my script still works... I still have other babel packages installed, but at another level in my project to be used with Lerna at compile time, so they should not take part in the execution of my scripts at runtime (and the fact that the delay is gone seems to confirm that "something" isn't executed anymore).

With all that in mind, would you be able to clarify these points ? :

1. What exactly is required to use Babel ?
1. Why ES6 still seems to work after removing `@babel/register` package entirely and the `compilers` Mocha option ?

Sorry for still no providing a reproducible example, but after this investigation, my questions are more generic this time, so I'd like to see what you guys think before taking more time setting something up. If you don't see immediately what could be wrong based on this new information though, just let me know and I will try to share a reproducible example this time.

Thank you for your help.

**Comment by BorisOsipov** [comment]
@nicholasbailey what is your nodejs version?

ES6 supported out of the box in modern nodejs versions(not all but most). You can check on https://node.green/

**Comment by nicolasbouffard** [comment]
That's strange. Currently we have nodejs `12.18.4`, but back when we included Babel, we had `11.3.0`, which shows up all "green" in the table you shared, although I'm sure that ES6 didn't work in our scripts back then (we also were on WebdriverIO 5 instead of 6).

So according to you we don't need to use Babel at all today ? If that's the case, why does the documentation sound like that's still a mandatory thing to be able to use ES6 ? Or maybe I understood it wrong ?

**Comment by BorisOsipov** [comment]
>So according to you we don't need to use Babel at all today ?
>If that's the case, why does the documentation sound like that's still a mandatory thing to be able to use ES6 ?

For example you have to use babel if you want ES6 import\export in nodejs<13. Many of users want(or not want but just coppied code from SO) use import\export. To make onbording seamless we propose babel. If you dont need you can delete it.

**Comment by christian-bromann** [comment]
I will close this issue as no reproducible example could be provided that indicates an issue with the stub protocol instance. I added a PR to find out the execution times of hooks.

**Comment by nicolasbouffard** [comment]
Sorry for the long silence and for re-opening the issue. I just wanted to say thanks and confirm that I can now use ES6 import syntax in my script without having to resort to babel. I just had to use `.mjs` as an extension for my scripts instead of `.js`.

Now my only problem is that I cannot use the `require` syntax anymore (I have legacy scripts that would require to be migrated) because WDIO throws a `Error: ReferenceError: require is not defined`. I found some information about [`createRequire`](https://nodejs.org/api/module.html#module_module_createrequire_filename) but couldn't make it work in my config file because of `SyntaxError: Cannot use 'import.meta' outside a module`. But that's a whole other issue so I don't expect you to answer here.
